File: src/main/java/org/dama/datasynth/matching/StochasticBlockModel.java
package org.dama.datasynth.matching;

import java.util.Arrays;
import java.util.Map;
import java.util.TreeMap;

/**
 * Created by aprat on 10/03/17.
 */
public class StochasticBlockModel <T> {

    private double [][] probabilities = null;
    private long [] sizes = null;
    private Map<T,Integer> mapping = null;

    public StochasticBlockModel(Map<T,Integer> mapping, long [] sizes, double [][] probabilities) {
        if(probabilities.length != sizes.length) throw new RuntimeException("sizes vector and probabilities matrix side missmatch");
        this.probabilities = probabilities;
        this.sizes = sizes;
        for(Map.Entry<T,Integer> entry : mapping.entrySet()) {
            if(entry.getValue() >= sizes.length) {
                throw new RuntimeException("Mapping between values and indices exceeds range");
            }
        }
        this.mapping = mapping;
    }

    public long getNumBlocks() {
        return sizes.length;
    }

    public long getSize(T value) {
        return sizes[mapping.get(value)];
    }

    public double getProbability( T i, T j) {
        int mapi = mapping.get(i);
        int mapj = mapping.get(j);
        return mapi < mapj ? probabilities[mapi][mapj] : probabilities[mapj][mapi];
    }

    public static <T extends Comparable<T>> StochasticBlockModel<T> extractFrom( long numEdges, Table<Long,T> values,
                                                                                 JointDistribution<T,T> distribution ) {

        Index<T> index = new Index(values);
        Map<T,Integer> mapping = new TreeMap<>();
        long [] sizes = new long[index.values().size()];
        int i = 0;
        for(T value : index.values()) {
            mapping.put(value,i);
            sizes[i] = index.getIds(value).size();
            i+=1;
        }

        double [][] probabilities = new double[sizes.length][sizes.length];
        for(i = 0; i < sizes.length; i+=1) {
            Arrays.fill(probabilities[i], 0.0D);
        }
        for(T value1 : index.values()) {
            i = mapping.get(value1);
            for( T value2 : index.values()) {
                int j = mapping.get(value2);
                if(i == j) {
                    probabilities[i][j] = 2.0D*numEdges*distribution.getProbability(new Tuple(value1,value2))/(sizes[i]*(sizes[j]-1));
                } else if( i < j) {
                    if(value1.compareTo(value2) < 0) {
                        probabilities[i][j] = numEdges * distribution.getProbability(new Tuple(value1, value2)) / (sizes[i] * (sizes[j]));
                    } else {
                        probabilities[i][j] = numEdges * distribution.getProbability(new Tuple(value2, value1)) / (sizes[i] * (sizes[j]));
                    }
                }
            }
        }
        return new StochasticBlockModel<T>(mapping,sizes,probabilities);
    }
}

File: src/main/scala/org/dama/datasynth/matching/utils/StochasticBlockModel.scala
package org.dama.datasynth.matching.utils

import scala.collection.mutable

/**
  * Created by joangui on 10/07/2017.
  */

object StochasticBlockModel
{
  def apply[T](probabilities:Array[Array[Double]],sizes:Array[Long],mapping:mutable.HashMap[T,Int]): StochasticBlockModel[T] ={
    if (probabilities.length != sizes.length)
      throw new RuntimeException("The stochastic block model is wrong, sizes vector and probabilities matrix side missmatch.")

    mapping.values.foreach(
      {case value => if(value>sizes.length)
        throw new RuntimeException("The stochastic block model is wrong, mapping between values and indices exceeds range.")
      })

    new StochasticBlockModel(probabilities,sizes,mapping)


  }
}


  class StochasticBlockModel[T] private (probabilities:Array[Array[Double]],sizes:Array[Long],mapping:mutable.HashMap[T,Int]){

  def getNumBlocks():Int = sizes.length

  def getSize(value:T):Long = {
    val index:Int=mapping.get(value).getOrElse(-1)
    if (index == -1)
      throw new RuntimeException("The stochastic block model is wrong. THIS SHOULD NEVER HAPPEN DUE TO PREVIOUS CHECKS.")
    sizes(index)
  }
    def getProbability(i:T,j:T):Double={
      val mapI:Int = mapping.get(i).getOrElse(-1)
      val mapJ:Int = mapping.get(j).getOrElse(-1)

       if (mapI < mapJ)
         probabilities(mapI)(mapJ)
       else
         probabilities(mapJ)(mapI)
    }

    def getMapping():Array[Array[Double]] = probabilities
    def getSizes():Array[Long] = sizes
}
